### PR TITLE
wrap UnsafeCell<MmapMut> with a mutex to fix Clippy lint on master

### PR DESCRIPTION
**Pull request by ylgrgyq**
According to my understanding, I tried to fix the problem mentioned here https://github.com/qdrant/wal/pull/56.

Please have a look and let me known if it is appropriate. Thanks

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -24,6 +24,7 @@ fs4 = "0.6"
 # to use direct functions on FreeBSD
 rustix = { version = "0", features = [ "fs" ]}
 crossbeam-channel = "0.5"
+parking_lot = { version = "0.12.1", features=["deadlock_detection"] }
```
Comment (agourlay) [inline Cargo.toml:27]: I don't think we should enable `deadlock_detection` by default at the library level.
This is a decision for the application developer I believe.

File: src/mmap_view_sync.rs [hunk 2]
```diff
@@ -13,7 +14,7 @@ use std::sync::Arc;
 /// The view may be split into disjoint ranges, each of which will share the
 /// underlying memory map.
 pub struct MmapViewSync {
-    inner: Arc<UnsafeCell<MmapMut>>,
+    inner: Arc<Mutex<UnsafeCell<MmapMut>>>,
```
Comment (agourlay) [inline src/mmap_view_sync.rs:17]: I wonder if we could actually get rid of the `UnsafeCell` completely now that a `Mutex` protects from concurrent access?

File: Cargo.toml (hunk 1 above)
Comment (ylgrgyq) [reply Cargo.toml]: Thanks. I fixed it

File: src/mmap_view_sync.rs (hunk 2 above)
Comment (ylgrgyq) [reply src/mmap_view_sync.rs:17]: I tried to remove `UnsafeCell` completely, but in `fn inner(&self) -> &MmapMut` and `fn inner_mut(&self) -> &mut MmapMut` the compiler says `cannot return value referencing temporary value`. It seems we still need this `UnsafeCell`.

**Comment by timvisee** [comment]
@ylgrgyq Thanks for giving this a shot! This is a tricky problem.

After some consideration I decided to bypass the warning for now (<https://github.com/qdrant/wal/pull/61>). Using a `Mutex` is not an option here as it would prevent having concurrent non-overlapping slice references.

I carefully inspected the code and do think all of it is sound in the current implementation.

I do believe that we should improve the architecture of this though to make it a lot harder to misuse.

**Pull request closed without merging**